### PR TITLE
Doc total order requirement of sort(_unstable)_by

### PR DESCRIPTION
**Pull request by Havvy**
I took the definition of what a total order is from the Ord trait
docs. I specifically put "elements of the slice" because if you
have a slice of f64s, but know none are NaN, then sorting by
partial ord is total in this case. I'm not sure if I should give
such an example in the docs or not.

r? @GuillaumeGomez

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Examples are always a big plus so please add them. :)

**Comment by Havvy** [comment]
Alright. I'll write them tomorrow. Sleep beckons.

**Comment by Havvy** [comment]
Tomorrow turned into a week. Whoops.

File: src/liballoc/slice.rs [hunk 1]
```diff
@@ -218,6 +218,15 @@ impl<T> [T] {
     /// * total and antisymmetric: exactly one of a < b, a == b or a > b is true; and
     /// * transitive, a < b and b < c implies a < c. The same must hold for both == and >.
     ///
+    /// For example, while `f64` doesn't implement `Ord` because `NaN != NaN`, we can use
```
Comment (GuillaumeGomez) [inline src/liballoc/slice.rs]: Please add urls to types.

File: src/libcore/slice/mod.rs [hunk 2]
```diff
@@ -1350,6 +1350,15 @@ impl<T> [T] {
     /// * total and antisymmetric: exactly one of a < b, a == b or a > b is true; and
     /// * transitive, a < b and b < c implies a < c. The same must hold for both == and >.
     ///
+    /// For example, while `f64` doesn't implement `Ord` because `NaN != NaN`, we can use
```
Comment (GuillaumeGomez) [inline src/libcore/slice/mod.rs]: Same

**Review by GuillaumeGomez — CHANGES_REQUESTED** [review]

**Comment by Havvy** [comment]
ping @GuillaumeGomez

File: src/liballoc/slice.rs [hunk 3]
```diff
@@ -211,6 +211,22 @@ impl<T> [T] {
...
     ///
+    /// The comparator function must define a total ordering for the elements in the slice. If
+    /// the ordering is not total, the order of the elements is unspecified. An order is a
+    /// total order if it is (for all a, b and c):
+    ///
+    /// * total and antisymmetric: exactly one of a < b, a == b or a > b is true; and
+    /// * transitive, a < b and b < c implies a < c. The same must hold for both == and >.
+    ///
+    /// For example, while [`f64`] doesn't implement [`Ord`] because `NaN != NaN`, we can use
+    /// `partial_cmp` as our sort function when we know the slice doesn't contain a `NaN`.
```
Comment (GuillaumeGomez) [inline src/liballoc/slice.rs]: Can you link to `partial_cmp` please?

File: src/libcore/slice/mod.rs [hunk 4]
```diff
@@ -1338,6 +1338,22 @@ impl<T> [T] {
...
     ///
+    /// The comparator function must define a total ordering for the elements in the slice. If
+    /// the ordering is not total, the order of the elements is unspecified. An order is a
+    /// total order if it is (for all a, b and c):
+    ///
+    /// * total and antisymmetric: exactly one of a < b, a == b or a > b is true; and
+    /// * transitive, a < b and b < c implies a < c. The same must hold for both == and >.
+    ///
+    /// For example, while [`f64`] doesn't implement [`Ord`] because `NaN != NaN`, we can use
+    /// `partial_cmp` as our sort function when we know the slice doesn't contain a `NaN`.
```
Comment (GuillaumeGomez) [inline src/libcore/slice/mod.rs]: Same.

**Comment by GuillaumeGomez** [comment]
Once the links added, it's good for me.

cc @rust-lang/docs (in case someone would like to double check).

**Comment by QuietMisdreavus** [comment]
+1. Can you add the links @GuillaumeGomez requested?

**Comment by Havvy** [comment]
Links added and rebased against master.

**Comment by Havvy** [comment]
I'm confused. Those links worked locally. I'll just add the "../" to make linkchecker happy though, and perhaps I misread what happened on my end?

File: src/liballoc/slice.rs [hunk 5]
```diff
@@ -252,6 +252,8 @@ impl<T> [T] {
     /// v.sort_by(|a, b| b.cmp(a));
     /// assert!(v == [5, 4, 3, 2, 1]);
     /// ```
+    ///
+    /// [`partial_cmp`]: ../../std/cmp/trait.PartialOrd.html#tymethod.partial_cmp
```
Comment (GuillaumeGomez) [inline src/liballoc/slice.rs]: Just remove this.

File: src/libcore/slice/mod.rs [hunk 6]
```diff
@@ -1378,6 +1378,7 @@ impl<T> [T] {
     /// assert!(v == [5, 4, 3, 2, 1]);
     /// ```
     ///
+    /// [`partial_cmp`]: ../../std/cmp/trait.PartialOrd.html#tymethod.partial_cmp
```
Comment (GuillaumeGomez) [inline src/libcore/slice/mod.rs]: And remove this one too.

File: src/liballoc/slice.rs (hunk 5 above)
Comment (Havvy) [reply src/liballoc/slice.rs]: That causes the rustdoc output to say:

```
222 |     /// [`partial_cmp`] as our sort function when we know the slice doesn't contain a `NaN`.                                                                                                                   
    |          ^^^^^^^^^^^^^ cannot be resolved, ignoring 
```

**Comment by QuietMisdreavus** [comment]
That `partial_ord` cannot be linked to, either from directly linking to the file or via intra-doc links.

* The core issue with linking directly to the file is that these docs appear in two places: directly on slices (`std/primitive.slice.html`) and on Vec when documented in std (`std/vec/struct.Vec.html` - they're not in alloc because the slice docs aren't available there). You can't use different amounts of `../` to appease both places.
* You could try to use intra-doc links like @GuillaumeGomez suggests, but you need to link to the libcore version of `PartialOrd`, which may cause the eventual docs to link to core instead of std like we prefer. If you want to try it, link to the method via the full declaration of the `PartialOrd` trait - so `core::cmp::PartialOrd::partial_cmp` (`cmp::PartialOrd::partial_cmp` for the location in libcore) - then see whether the docs in std link to core or std. I have a feeling they'll link to core, but i haven't looked closely enough at the relevant code to know for sure.

**Comment by TimNN** [comment]
ping from triage @Havvy: it looks like some changes have been requested to this PR.

**Comment by TimNN** [comment]
Ping from triage @GuillaumeGomez: it looks like this PR is ready for another review.

**Comment by TimNN** [comment]
Ping from triage @GuillaumeGomez / @rust-lang/docs: This PR requires your review.

**Comment by GuillaumeGomez** [comment]
Then it's all good as is. Thanks!

@bors: r+

**Pull request merged**